typescript/redis_work_queue/index.ts

```diff
@@ -0,0 +1,251 @@
+    /**
+     * Add an item to the work queue. This adds the redis commands onto the pipeline passed.
+     * Use `WorkQueue.addItem` if you don't want to pass a pipeline directly.
+     * Add the item data
+     * NOTE: it's important that the data is added first, otherwise someone before the data is
+     * ready
+     */
+    const itemId = item.Id();
+    pipeline.set(this.itemDataKey + itemId, item.Data());
+    // Then add the id to the work queue
+    pipeline.lpush(this.mainQueueKey, itemId);
+  }
+
+  addItem(db: Redis, item: typeof Item): void {
+    // Add an item to the work queue.
+    // This creates a pipeline and executes it on the database.
+    const pipeline = db.pipeline() as unknown as Pipeline;
+    this.addItemToPipeline(pipeline, item);
+    pipeline.exec();
+  }
+
+  queueLen(db: Redis): Promise<number> {
+    // Return the length of the work queue (not including items being processed, see `WorkQueue.processing()`).
+    return db.llen(this.mainQueueKey);
+  }
+
+  processing(db: Redis) {
+    // Return the number of items being processed.
+    return db.llen(this.processingKey);
+  }
+
+  _lease_exists(db: Redis, itemId: string): Promise<boolean> {
+    return db.exists(KeyPrefix(this.leaseKey).of(itemId)).then(exists => exists !== 0);
+  }
+  
+
+  async lease(db: Redis, leaseSecs: number, block = true, timeout = 0): Promise<typeof Item> {
+    /**
+     * Request a work lease the work queue. This should be called by a worker to get work to
+     * complete. When completed, the `complete` method should be called.
+     *
+     * If `block` is true, the function will return either when a job is leased or after `timeout`
+     * if `timeout` isn't 0.
+     *
+     * If the job is not completed before the end of `lease_duration`, another worker may pick up
+     * the same job. It is not a problem if a job is marked as `done` more than once.
+     *
+     * If you've not already done it, it's worth reading [the documentation on leasing
+     * items](https://github.com/MeVitae/redis-work-queue/blob/main/README.md#leasing-an-item).
+     */
+    let maybeItemId: Buffer | string | null;
+    let itemId: string;
+    // First, to get an item, we try to move an item from the main queue to the processing list.
+    if (block) {
+      maybeItemId = await db.brpoplpush(this.mainQueueKey, this.processingKey, timeout);
+    } else {
+      maybeItemId = await db.rpoplpush(this.mainQueueKey, this.processingKey);
+    }
+
+    if (maybeItemId == null) {
+      return undefined;
+    }
+
+    // Make sure the item id is a string
+    if (Buffer.isBuffer(maybeItemId)) {
+      itemId = maybeItemId.toString('utf-8');
+    } else if (typeof maybeItemId === 'string') {
+      itemId = maybeItemId;
+    } else {
+      throw new Error("item id from work queue not bytes or string");
+    }
+
+    // If we got an item, fetch the associated data.
+    let data: Buffer | string | null = await db.get(KeyPrefix(this.itemDataKey).of(itemId));
```

Comment:
What's the behaviour of `get`, what happens if the value can't be converted to a string?